File: app/src/main/java/com/wxm/camerajob/utility/FileLogger.java
package com.wxm.camerajob.utility;

import android.os.Environment;
import android.support.design.BuildConfig;

import java.io.File;
import java.io.IOException;
import java.util.Locale;
import java.util.logging.FileHandler;
import java.util.logging.Level;
import java.util.logging.LogRecord;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

import wxm.androidutil.util.UtilFun;

/**
 * 写在文件中的日志
 * Created by 123 on 2016/6/18.
 */
@SuppressWarnings("ResultOfMethodCallIgnored")
public class FileLogger {
    private final static String LOG_NAME = "camerajob_run_%g.log";

    private String mLogTag;
    private Logger mLoger;

    private static FileLogger instance;

    private static FileLogger getInstance() {
        if (null == instance) {
            instance = new FileLogger();
        }

        return instance;
    }

    public static Logger getLogger() {
        return getInstance().mLoger;
    }

    private FileLogger() {
        mLogTag =  ("P" + System.currentTimeMillis() % 100000);

        String logfn;
        String en = Environment.getExternalStorageState();
        if (en.equals(Environment.MEDIA_MOUNTED)) {
            File sdcardDir = Environment.getExternalStorageDirectory();
            String path = sdcardDir.getPath() + "/CamerajobLogs";
            File path1 = new File(path);
            if (!path1.exists()) {
                path1.mkdirs();
            }

            logfn = path + "/" + LOG_NAME;

        } else {
            File innerPath = ContextUtil.getInstance()
                    .getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS);
            assert innerPath != null;
            logfn = innerPath.getPath() + "/" + LOG_NAME;
        }

        try {
            FileHandler mLogFH = new FileHandler(logfn, true);
            mLogFH.setFormatter(new SimpleFormatter() {
                @Override
                public String format(LogRecord record) {

                    String ret = String.format(Locale.CHINA, "%s|%s|%s-%d|%s:%s|%s"
                            , UtilFun.MilliSecsToString(record.getMillis())
                            , record.getLevel().getName()
                            , mLogTag ,record.getThreadID()
                            , record.getSourceClassName(), record.getSourceMethodName()
                            , formatMessage(record)) + (System.lineSeparator());
                    return ret;
                }
            });

            mLoger = Logger.getLogger("camerajob_runlog");
            mLoger.addHandler(mLogFH);

            if(BuildConfig.DEBUG)
                mLoger.setLevel(Level.INFO);
            else
                mLoger.setLevel(Level.WARNING);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/wxm/camerajob/utility/FileLogger.kt
package com.wxm.camerajob.utility

import android.os.Environment
import android.support.design.BuildConfig
import android.util.Log
import wxm.androidutil.util.UtilFun
import java.io.File
import java.io.IOException
import java.util.*
import java.util.logging.*

/**
 * log in file
 * Created by WangXM on 2016/6/18.
 */
class FileLogger private constructor() {
    private val mLogTag: String = "P" + System.currentTimeMillis() % 100000
    private lateinit var mLogger: Logger

    init {
        val fn = String.format(Locale.CHINA, LOG_NAME, mLogTag)
        val logFileName = "${ContextUtil.getAppRootDir()}/$fn".apply{
            File(this).let {
                    if (!it.exists()) {
                        it.mkdirs()
                    }
                }
        }

        try {
            FileHandler(logFileName, true).let {
                it.formatter = object : SimpleFormatter() {
                    override fun format(record: LogRecord): String {
                        return String.format(Locale.CHINA,
                                "%s|%s|%s-%d|%s:%s|%s",
                                UtilFun.MilliSecsToString(record.millis),
                                record.level.name,
                                mLogTag, record.threadID,
                                record.sourceClassName,
                                record.sourceMethodName,
                                formatMessage(record)) + System.lineSeparator()
                    }
                }

                mLogger = Logger.getLogger("cameraJobRunLog")
                mLogger.addHandler(it)
            }

            @Suppress("ConstantConditionIf")
            mLogger.level = if (BuildConfig.DEBUG) Level.INFO else Level.WARNING
        } catch (e: IOException) {
            Log.e(LOG_TAG, "create file for log failure", e)
        }
    }

    companion object {
        private val LOG_TAG = ::FileLogger.javaClass.simpleName
        private const val LOG_NAME = "cameraJob_run_%s.log"

        private val instance: FileLogger = FileLogger()
        var logger: Logger = instance.mLogger
    }
}
